### ISSUE TITLE
improvement suggestion: change keyboard layout

### ISSUE THREAD

**Issue by logg-sar** [biosbits/bits#2]
Hi Josh

first of all - thanks for your great work!

Unfortunately i´m using an different keyboard layout.
It is possible, but quite annoying to search the right character when its needed.

Therefore i would like to ask you if you could create a set of gkb-files with the most used keyboard layouts. In my case german ;)
It seems that this gkb files have to be created _exactly_ with the correct grub version. Else there is only an error message.

Thanks
Netlord

**Comment by joshtriplett** [comment]
It looks like such files get generated by either grub-mklayout (using Linux keymaps) or grub-kbdcomp (from X keyboard layouts).

It also looks like it suffices to use that tool with the same GRUB_KEYBOARD_LAYOUTS_VERSION, but that has indeed changed from 8 to 10 between our BITS snapshot and current upstream GRUB2.  As a temporary workaround, try building a layout file using grub-mklayout or grub-kbdcomp built from the same version of GRUB2 that BITS uses.  Can you confirm that doing so works for you?

I'm hesitant to have BITS build-depend on Linux or X keyboard map files and attempt to generate all possible keyboard maps people might want.  (Especially in the latter case, which could include arbitrary combinations of keyboard maps and options.)  What I'm tempted to do, instead:

1) Upgrade to current GRUB2.  We need to do this anyway, for many reasons.  That will mean that you can build keyboard layouts using the GRUB2 that ships with your distribution.

2) Add support to automatically handle keyboard layout files placed in a specific location on a BITS image.  (For instance, showing a menu of available layouts, and automatically loading a "default" keyboard layout.)

3) Document (1) and (2) so that anyone who needs another keyboard layout can build and include one such that BITS will use it.

Would that work for you?

**Comment by logg-sar** [comment]
My main system is a centos7 with an GRUB2 in the version 2.02-0.17.el7-1.4

grub-kbdcomp is a wrapper-script for grub-mklayout. I´ve created an gkb-file but when i´m using it i get the error "invalid version".

when I try to update bits with the files from centos i get while booting the error "symbol 'grub_file_progress_hook' not found"

can you tell me _exactly_ which version of grub you use?`
i would install this and create an new gkb file

thanks

**Comment by joshtriplett** [comment]
> My main system is a centos7 with an GRUB2 in the version 2.02-0.17.el7-1.4
> grub-kbdcomp is a wrapper-script for grub-mklayout. I´ve created an gkb-file but when i´m using it i get the error "invalid version".

Right; 2.02 uses GRUB_KEYBOARD_LAYOUTS_VERSION 10.  We use a modified 2.00, which uses GRUB_KEYBOARD_LAYOUTS_VERSION 8.

> when I try to update bits with the files from centos i get while booting the error "symbol 'grub_file_progress_hook' not found"

Yeah, that won't work.  You can't mix modules from your system GRUB with BITS.

> can you tell me exactly which version of grub you use?`
> i would install this and create an new gkb file

Modified 2.00.  Try cloning the sources from https://github.com/biosbits/grub, then run ./autogen.sh, ./configure --prefix=/tmp/grub-inst , make install, and then run grub-mklayout from that installation path.
